src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+  console.log(result.value)
+  result = countdown.next()
+}
+```
+
+This will produce the following output:
+
+```text
+3
+2
+1
+0
+```
+
+This is not the most intuitive or ergonomic approach, but the iterator protocol is the basic building block for the _iterable protocol_ which enables the `for...of` syntax.
+
+## The JavaScript iterable protocol
+
+As we said, the _iterable protocol_ builds on top of the _iterator protocol_ that we just explored. Let's see how:
+
+> An object is **iterable** if it implements the `@@iterator` method, a zero-argument function that **returns an iterator**.
+
+Note that with `@@iterator` we indicate a symbol that is accessed with the global value `Symbol.iterator`.
+
+Can we make our countdown example _iterable_? We certainly can!
+
+```javascript
+function createCountdown (from) {
+  let nextVal = from
+  return {
+    [Symbol.iterator]: () => ({
+      next () {
+        if (nextVal < 0) {
+          return { done: true }
+        }
+
+        return { done: false, value: nextVal-- }
+      }
+    })
+  }
+}
+```
+
+In this new example, our factory function returns an _iterable_ object. This object in fact has a method referenced with `Symbol.iterator` that returns an _iterator_ object.
+
+At this point, once we have an instance of a countdown, we can use the `for..of` syntax to iterate over all the numbers from the countdown:
+
+```javascript
+const countdown = createCountdown(3)
+
+for (const value of countdown) {
+  console.log(value)
+}
+```
+
+This will output:
+
+```text
+3
+2
+1
+0
+```
+
+Hooray! Now we know how to make iterators and iterable objects. If you find the two terms confusing, don't worry, that's quite common!
+
+One good way to try to remember and distinguish these 2 concepts is the following:
+
+  - An _iterator_ is a lower-level object that allows us to iterate over some data set using `next()`
+  - An _iterable_ is an object on which we can iterate over using the `for...of` syntax.
+
+
+### Using JavaScript generator functions
+
+An interesting detail is that JavaScript generator functions _produce_ iterators.
+
+This allows us to simplify the way we can implement both the _iterator_ and the _iterable_ protocols.
+
+Let's see how can we rewrite our countdown logic using a generator function:
+
+```javascript
+function * createCountdown (from) {
+  for (let i = from; i >= 0; i--) {
+    yield i
+  }
+}
+```
+
+If we call `createCountdown(3)` we get an _iterator_. So this is perfectly compatible with our previous _iterator_ implementation:
+
+```javascript
+const countdown = createCountdown(3)
+console.log(countdown.next()) // { value: 3, done: false }
+console.log(countdown.next()) // { value: 2, done: false }
+console.log(countdown.next()) // { value: 1, done: false }
+console.log(countdown.next()) // { value: 0, done: false }
+console.log(countdown.next()) // { value: undefined, done: true }
+```
+
+Similarly, we can use generators to implement the _iterable protocol_:
```

Comment:
No need for this. The object returned by a generator is already iterable as it is.